tools/cli/main.cpp

```diff
@@ -12,246 +12,859 @@
+    }
+
+    if (in_count == 1 && is_hyphen(std::string(in_fnames[0]))) {
+        common_args->input_files.emplace_back(IonCliIO("stdin", IO_TYPE_CONSOLE));
+    } else if (!ion_cli_input_names_clean(in_count, in_fnames)) {
+        iFAIL_MSG(IERR_INVALID_ARG, "Input can be '-' or file name(s) but not both.")
+    } else {
+        common_args->input_files = ion_cli_arg_names_to_io(in_count, in_fnames);
     }
 
-    IONREPORT(ion_cli_args_common(&args, &common_args, &report));
-    if (ion_cli_has_flag(&args, "process")) {
-        IonCliProcessArgs process_args;
-        IONREPORT(ion_cli_args_process(&args, &process_args));
-        IONREPORT(ion_cli_command_process(&common_args, &process_args, NULL, &report));
+    cleanup:
+    if (result->has_error_description) {
+        report->addResult(result);
     }
-    else if (ion_cli_has_flag(&args, "compare")) {
-        ION_EVENT_COMPARISON_TYPE comparison_type = COMPARISON_TYPE_BASIC;
-        if (ion_cli_has_value(&args, "--comparison-type")) {
-            comparison_type = ion_cli_comparison_type_from_string(args.find("--comparison-type")->second.asString());
+    iRETURN;
+}
+
+iERR ion_help_subcommand(int sub_command, const char **sub_cmd_name) {
+    iENTER;
+
+    if (sub_command > 1) {
+        std::cerr << "Help subcommand accepts only one of process, compare, extract." << std::endl;
+        iFAIL(IERR_INVALID_ARG)
+    } else if (sub_command == 0) {
+        ion_print_full_help_message();
+        SUCCEED()
+    } else {
+        std::string sname = std::string(*sub_cmd_name);
+        if (sname == "process") {
+            PRINT_SUBCOMMAND_HELP_ON_ERROR(proc_argtable)
+            ion_print_process_examples();
+            SUCCEED()
+        }
+        if (sname == "compare") {
+            PRINT_SUBCOMMAND_HELP_ON_ERROR(comp_argtable)
+            ion_print_compare_examples();
+            SUCCEED()
+        }
+        if (sname == "extract") {
+            PRINT_SUBCOMMAND_HELP_ON_ERROR(xtract_argtable)
+            ion_print_extract_examples();
+            SUCCEED()
         }
-        IONREPORT(ion_cli_command_compare(&common_args, comparison_type, NULL, &report));
     }
-    else if (ion_cli_has_flag(&args, "extract")) {
-        // TODO implement extract
-        std::cerr << "extract command not yet implemented" << std::endl;
+    iRETURN;
+}
+
+/**
+ * Used for when a subcommand is given some valid and some invalid inputs.
+ * Shorter message (no examples) than the message printed with `help` subcommand.
+ */
+void print_usage_for_subcommand(struct arg_end *table_end, void **args_table) {
+    arg_print_errors(stdout, table_end, ION_CLI_PNAME);
+    std::cout << std::endl << "Usage : " << std::endl;
+    PRINT_ION_COMMAND;
+    arg_print_syntaxv(stdout, args_table, "\n");
+    arg_print_glossary_gnu(stdout, args_table);
+}
+
+/**
+ * Check that each argtable instance was created and initialized.
+ */
+bool arg_tables_initalized() {
```

Comment:
Based on usage, it looks like this should be named something like `arg_tables_not_initialized`.